gobblin-runtime/src/main/java/gobblin/runtime/app/ServiceBasedAppLauncher.java

```diff
@@ -0,0 +1,203 @@
+ * </ul>
+ *
+ * <p>
+ *   Additional {@link Service}s can be added via the {@link #addService(Service)} method. A {@link Service} cannot be
+ *   added after the application has started.
+ * </p>
+ */
+public class ServiceBasedAppLauncher implements ApplicationLauncher {
+
+  /**
+   * The name of the application. Not applicable for YARN jobs, which uses a separate key for the application name.
+   */
+  public static final String APP_NAME = "app.name";
+
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceBasedAppLauncher.class);
+
+  private final String appId;
+  private final List<Service> services;
+
+  private volatile boolean hasStarted = false;
+  private volatile boolean hasStopped = false;
+
+  private ServiceManager serviceManager;
+
+  public ServiceBasedAppLauncher(Properties properties, String appName) throws Exception {
+    this.appId = ApplicationLauncherUtils.newAppId(appName);
+    this.services = new ArrayList<>();
+
+    addJobExecutionServerAndAdminUI(properties);
+    addMetricsService(properties);
+    addJMXReportingService();
+
+    // Add a shutdown hook that interrupts the main thread
+    addInterruptedShutdownHook();
+  }
+
+  /**
+   * Starts the {@link ApplicationLauncher} by starting all associated services. This method also adds a shutdown hook
+   * that invokes {@link #stop()} and the {@link #close()} methods. So {@link #stop()} and {@link #close()} need not be
+   * called explicitly; they can be triggered during the JVM shutdown.
+   */
+  @Override
+  public void start() {
+    if (this.hasStarted) {
+      LOG.warn("ApplicationLauncher has already started");
+      return;
+    } else {
+      this.hasStarted = true;
+    }
+
+    this.serviceManager = new ServiceManager(this.services);
+
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      public void run() {
+        try {
+          ServiceBasedAppLauncher.this.stop();
+        } catch (ApplicationException e) {
+          LOG.error("Failed to shutdown application", e);
+        } finally {
+          try {
+            ServiceBasedAppLauncher.this.close();
+          } catch (IOException e) {
+            LOG.error("Failed to close application", e);
+          }
+        }
+      }
+    });
+
+    LOG.info("Starting the Gobblin application and all its associated Services");
+
+    // Start the application
+    this.serviceManager.startAsync();
+  }
+
+  /**
+   * Stops the {@link ApplicationLauncher} by stopping all associated services.
+   */
+  @Override
+  public void stop() throws ApplicationException {
+    if (this.hasStopped) {
+      LOG.warn("ApplicationLauncher has already stopped");
+      return;
+    } else {
+      this.hasStopped = true;
+    }
+
+    LOG.info("Shutting down the application");
+    try {
+      // Give the services 5 seconds to stop to ensure that we are responsive to shutdown requests
+      this.serviceManager.stopAsync().awaitStopped(5, TimeUnit.SECONDS);
```

Comment:
5 seconds seems really short.  The `GobblinApplicationMaster` used to wait for 5 minutes.